File: .conflict-base-0/app/src/test/java/org/example/cert/CertificateValidatorTest.java
package org.example.cert;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.File;
import java.io.FileOutputStream;
import java.net.IDN;
import java.security.KeyPair;
import java.security.KeyStore;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class CertificateValidatorTest {
    private CertificateValidator validator;
    private File keystoreFile;
    private static final String KEYSTORE_PASSWORD = "password";
    private KeyPair keyPair;
    private X509Certificate testCertificate;
    private X509Certificate ipCertificate;
    private X509Certificate multiDomainCertificate;

    @TempDir
    File tempDir;

    @BeforeEach
    void setUp() throws Exception {
        // Create a test keystore
        keystoreFile = new File(tempDir, "test.jks");
        createTestKeystore();
        validator = new CertificateValidator(keystoreFile, KEYSTORE_PASSWORD);

        // Generate test certificate with standard domains
        keyPair = TestCertificateGenerator.generateKeyPair();
        testCertificate = TestCertificateGenerator.generateCertificate(
            "CN=example.com, O=Test Organization, C=US",
            "CN=Test CA, O=Test Organization, C=US",
            keyPair,
            keyPair
        );

        // Generate certificate with IP addresses
        ipCertificate = TestCertificateGenerator.generateCertificate(
            "CN=192.168.1.1, O=Test Organization, C=US",
            "CN=Test CA, O=Test Organization, C=US",
            keyPair,
            keyPair,
            null,
            new String[]{"192.168.1.1", "10.0.0.1"}
        );

        // Generate certificate with multiple domains including IDN
        multiDomainCertificate = TestCertificateGenerator.generateCertificate(
            "CN=multi.example.com, O=Test Organization, C=US",
            "CN=Test CA, O=Test Organization, C=US",
            keyPair,
            keyPair,
            new String[]{"xn--80akhbyknj4f.xn--p1ai", "subdomain.example.com", "*.test.example.com"},
            null
        );

        // Add the certificates to the keystore
        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
        ks.load(keystoreFile.toURI().toURL().openStream(), KEYSTORE_PASSWORD.toCharArray());
        ks.setCertificateEntry("test-cert", testCertificate);
        ks.setCertificateEntry("ip-cert", ipCertificate);
        ks.setCertificateEntry("multi-domain-cert", multiDomainCertificate);
        try (FileOutputStream fos = new FileOutputStream(keystoreFile)) {
            ks.store(fos, KEYSTORE_PASSWORD.toCharArray());
        }
    }

    private void createTestKeystore() throws Exception {
        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
        ks.load(null, null);
        try (FileOutputStream fos = new FileOutputStream(keystoreFile)) {
            ks.store(fos, KEYSTORE_PASSWORD.toCharArray());
        }
    }

    @Test
    void testValidateCertificateChain_WithValidCertificates() throws Exception {
        // Create test certificates
        Certificate[] certs = new Certificate[]{testCertificate};

        // Test validation
        X509Certificate[] validatedCerts = validator.validateCertificateChain(certs);

        assertNotNull(validatedCerts);
        assertEquals(certs.length, validatedCerts.length);
    }

    @Test
    void testValidateCertificateChain_WithInvalidCertificates() {
        // Create invalid certificates
        Certificate[] invalidCerts = new Certificate[0];

        // Test validation
        assertThrows(CertificateException.class, () ->
            validator.validateCertificateChain(invalidCerts)
        );
    }

    @Test
    void testGetCertificateInfo() throws Exception {
        // Get certificate info
        Map<String, Object> certInfo = validator.getCertificateInfo(testCertificate);

        // Verify the info
        assertNotNull(certInfo);
        assertTrue(certInfo.containsKey("subjectDN"));
        assertTrue(certInfo.containsKey("issuerDN"));
        assertTrue(certInfo.containsKey("version"));
        assertTrue(certInfo.containsKey("serialNumber"));
        assertTrue(certInfo.containsKey("validFrom"));
        assertTrue(certInfo.containsKey("validUntil"));
        assertTrue(certInfo.containsKey("signatureAlgorithm"));
        assertTrue(certInfo.containsKey("publicKeyAlgorithm"));
    }

    @Test
    void testVerifyHostname_ExactMatch() throws Exception {
        assertTrue(validator.verifyHostname(testCertificate, "example.com"));
    }

    @Test
    void testVerifyHostname_WildcardMatch() throws Exception {
        // 基本通配符匹配
        assertTrue(validator.verifyHostname(testCertificate, "sub.example.com"));

        // 确保通配符只匹配一个级别
        assertFalse(validator.verifyHostname(testCertificate, "sub.sub.example.com"));
    }

    @Test
    void testVerifyHostname_NoMatch() throws Exception {
        assertFalse(validator.verifyHostname(testCertificate, "other.com"));
    }

    @Test
    void testVerifyHostname_WithSubjectAltNames() throws Exception {
        assertTrue(validator.verifyHostname(testCertificate, "alt.example.com"));
    }

    @Test
    void testVerifyHostname_WithIDN() throws Exception {
        // 输出证书中的SAN扩展内容
        System.out.println("\n========== IDN测试开始 ==========");
        Collection<List<?>> sans = multiDomainCertificate.getSubjectAlternativeNames();
        if (sans != null) {
            System.out.println("证书中的SubjectAlternativeNames:");
            for (var san : sans) {
                Integer type = (Integer) san.get(0);
                String value = (String) san.get(1);
                System.out.println("  类型: " + type + ", 值: " + value);
            }
        }

        // 获取证书中的IDN域名
        String certIdnValue = extractCertificateIDN(sans);
        System.out.println("证书中的IDN域名: " + certIdnValue);

        if (certIdnValue != null) {
            // 先测试Punycode格式直接匹配
            System.out.println("\n测试1: 直接使用证书中的Punycode进行验证");
            boolean punycodeResult = validator.verifyHostname(multiDomainCertificate, certIdnValue);
            System.out.println("Punycode直接匹配结果: " + punycodeResult);
            assertTrue(punycodeResult, "使用证书中的Punycode域名应该匹配成功");

            // 将Punycode转换为Unicode并测试
            String unicodeHostname = IDN.toUnicode(certIdnValue);
            System.out.println("\n测试2: 将证书中的Punycode转换为Unicode后验证");
            System.out.println("Punycode转Unicode: " + unicodeHostname);

            boolean unicodeResult = validator.verifyHostname(multiDomainCertificate, unicodeHostname);
            System.out.println("Unicode格式匹配结果: " + unicodeResult);
            assertTrue(unicodeResult, "转换后的Unicode域名应该匹配成功");
        } else {
            fail("证书中没有找到IDN域名");
        }

        System.out.println("========== IDN测试结束 ==========\n");
    }

    /**
     * 从证书的SAN扩展中提取第一个IDN域名
     */
    private String extractCertificateIDN(Collection<List<?>> sans) {
        if (sans == null) return null;

        for (var san : sans) {
            Integer type = (Integer) san.get(0);
            String value = (String) san.get(1);

            // DNS类型 = 2，并且是Punycode格式的IDN
            if (type == 2 && value.contains("xn--")) {
                return value;
            }
        }
        return null;
    }

    @Test
    void testVerifyHostname_WithIPAddress() throws Exception {
        // 验证IP地址匹配
        assertTrue(validator.verifyHostname(ipCertificate, "192.168.1.1"));
        assertTrue(validator.verifyHostname(ipCertificate, "10.0.0.1"));

        // 不匹配的IP地址
        assertFalse(validator.verifyHostname(ipCertificate, "192.168.1.2"));

        // IP地址不应匹配DNS名称
        assertFalse(validator.verifyHostname(testCertificate, "192.168.1.1"));
    }

    @Test
    void testVerifyHostname_EdgeCases() throws Exception {
        // 带有尾部点的域名
        assertTrue(validator.verifyHostname(testCertificate, "example.com."));

        // 大小写不敏感性测试
        assertTrue(validator.verifyHostname(testCertificate, "EXAMPLE.com"));

        // 多级域名匹配测试
        assertTrue(validator.verifyHostname(multiDomainCertificate, "subdomain.example.com"));

        // 通配符边界测试
        assertTrue(validator.verifyHostname(multiDomainCertificate, "www.test.example.com"));
        assertFalse(validator.verifyHostname(multiDomainCertificate, "www.other.example.com"));

        // 空主机名
        assertFalse(validator.verifyHostname(testCertificate, null));
        assertFalse(validator.verifyHostname(testCertificate, ""));
    }
}

File: app/src/test/kotlin/org/example/cert/CertificateValidatorTest.kt
package org.example.cert

import kotlin.test.*
import org.bouncycastle.asn1.DEROctetString
import java.io.File
import java.io.FileOutputStream
import java.math.BigInteger
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.time.Duration
import java.time.Instant
import java.util.*
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.asn1.x509.*
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter
import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder
import org.example.config.SSLTestConfig
import java.net.InetAddress

class CertificateValidatorTest {
    private lateinit var tempDir: File
    private var config: SSLTestConfig? = null
    private var validator: CertificateValidator? = null
    private lateinit var validKeyPair: KeyPair
    private lateinit var validCertificate: X509Certificate
    private lateinit var expiredCertificate: X509Certificate
    private lateinit var notYetValidCertificate: X509Certificate
    private lateinit var selfSignedCertificate: X509Certificate
    private lateinit var intermediateCertificate: X509Certificate
    private lateinit var leafCertificate: X509Certificate
    private lateinit var rootCertificate: X509Certificate

    @BeforeTest
    fun setup() {
        tempDir = createTempDir()
        // Generate test certificates
        validKeyPair = generateKeyPair()
        validCertificate = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("example.com", "*.example.com")
        )

        expiredCertificate = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=expired.example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(365)),
            validUntil = Instant.now().minus(Duration.ofDays(1)),
            domains = listOf("expired.example.com")
        )

        notYetValidCertificate = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=future.example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().plus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("future.example.com")
        )

        selfSignedCertificate = generateSelfSignedCertificate(
            keyPair = validKeyPair,
            subject = "CN=self-signed.example.com",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("self-signed.example.com")
        )

        // Create a chain: root CA -> intermediate CA -> leaf certificate
        val rootKeyPair = generateKeyPair()
        val intermediateKeyPair = generateKeyPair()
        val leafKeyPair = generateKeyPair()

        rootCertificate = generateSelfSignedCertificate(
            keyPair = rootKeyPair,
            subject = "CN=Root CA",
            validFrom = Instant.now().minus(Duration.ofDays(10)),
            validUntil = Instant.now().plus(Duration.ofDays(3650)),
            isCA = true
        )

        intermediateCertificate = generateCertificate(
            keyPair = intermediateKeyPair,
            subject = "CN=Intermediate CA",
            issuer = "CN=Root CA",
            validFrom = Instant.now().minus(Duration.ofDays(5)),
            validUntil = Instant.now().plus(Duration.ofDays(1825)),
            isCA = true,
            issuerKeyPair = rootKeyPair
        )

        leafCertificate = generateCertificate(
            keyPair = leafKeyPair,
            subject = "CN=leaf.example.com",
            issuer = "CN=Intermediate CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("leaf.example.com", "www.leaf.example.com"),
            issuerKeyPair = intermediateKeyPair
        )

        // Create a test keystore with the root certificate
        val keystoreFile = File(tempDir, "test-truststore.jks")
        val keystore = KeyStore.getInstance(KeyStore.getDefaultType())
        keystore.load(null, null)
        keystore.setCertificateEntry("root-ca", rootCertificate)
        keystore.setCertificateEntry("self-signed", selfSignedCertificate)

        FileOutputStream(keystoreFile).use { fos ->
            keystore.store(fos, "password".toCharArray())
        }

        // Configure the validator with a real SSLTestConfig instance
        val testConfig = SSLTestConfig()
        testConfig.keystoreFile = keystoreFile
        testConfig.keystorePassword = "password"
        config = testConfig
        validator = CertificateValidator(config!!)
    }

    @Test
    fun `validateCertificateChain should validate certificate chain`() {
        val result = validator!!.validateCertificateChain(
            arrayOf(leafCertificate, intermediateCertificate),
            "leaf.example.com"
        )
        assertEquals(2, result.size)
        assertSame(leafCertificate, result[0])
        assertSame(intermediateCertificate, result[1])
    }

    @Test
    fun `validateCertificateChain should throw exception for hostname mismatch`() {
        val exception = assertFailsWith<CertificateException> {
            validator!!.validateCertificateChain(
                arrayOf(leafCertificate, intermediateCertificate),
                "wrong.example.com"
            )
        }
        assertTrue(exception.message?.contains("Hostname verification failed") == true)
    }

    @Test
    fun `verifyHostname should match exact hostname`() {
        val cert = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=exact.example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("exact.example.com")
        )

        assertTrue(validator!!.verifyHostname(cert, "exact.example.com"))
        assertFalse(validator!!.verifyHostname(cert, "other.example.com"))
    }

    @Test
    fun `verifyHostname should match wildcard hostname`() {
        val cert = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=*.example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("*.example.com")
        )

        assertTrue(validator!!.verifyHostname(cert, "sub.example.com"))
        assertTrue(validator!!.verifyHostname(cert, "test.example.com"))
        assertFalse(validator!!.verifyHostname(cert, "example.com"))
        assertFalse(validator!!.verifyHostname(cert, "sub.sub.example.com"))
    }

    @Test
    fun `verifyHostname should match IP address`() {
        val cert = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=192.168.1.1",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            ipAddresses = listOf("192.168.1.1")
        )

        // Get the IP address from the certificate's SAN extension
        val sans = cert.getSubjectAlternativeNames()
        println("All SANs: $sans")
        val ipSan = sans?.find { it[0] == 7 }
        println("IP SAN: $ipSan")
        println("IP SAN type: ${ipSan?.get(1)?.javaClass?.name}")
        println("IP SAN value: ${ipSan?.get(1)}")

        assertTrue(validator!!.verifyHostname(cert, "192.168.1.1"))
        assertFalse(validator!!.verifyHostname(cert, "192.168.1.2"))
    }

    @Test
    fun `verifyHostname should match Common Name when no SAN present`() {
        val cert = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=cn.example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365))
        )

        assertTrue(validator!!.verifyHostname(cert, "cn.example.com"))
        assertFalse(validator!!.verifyHostname(cert, "other.example.com"))
    }

    @Test
    fun `verifyHostname should return false for empty hostname`() {
        val cert = generateCertificate(
            keyPair = validKeyPair,
            subject = "CN=example.com",
            issuer = "CN=Test CA",
            validFrom = Instant.now().minus(Duration.ofDays(1)),
            validUntil = Instant.now().plus(Duration.ofDays(365)),
            domains = listOf("example.com")
        )

        assertFalse(validator!!.verifyHostname(cert, ""))
    }

    @Test
    fun `checkClientTrusted should throw exception`() {
        val exception = assertFailsWith<CertificateException> {
            validator!!.checkClientTrusted(arrayOf(validCertificate), "RSA")
        }
        assertEquals("Client certificate validation not supported", exception.message)
    }

    @Test
    fun `checkServerTrusted should throw exception for null chain`() {
        val exception = assertFailsWith<CertificateException> {
            validator!!.checkServerTrusted(null, "RSA")
        }
        assertEquals("Certificate chain is null or empty", exception.message)
    }

    @Test
    fun `checkServerTrusted should throw exception for empty chain`() {
        val exception = assertFailsWith<CertificateException> {
            validator!!.checkServerTrusted(emptyArray(), "RSA")
        }
        assertEquals("Certificate chain is null or empty", exception.message)
    }

    // Helper functions for certificate generation
    private fun generateKeyPair(): KeyPair {
        val generator = KeyPairGenerator.getInstance("RSA")
        generator.initialize(2048)
        return generator.generateKeyPair()
    }

    private fun generateCertificate(
        keyPair: KeyPair,
        subject: String,
        issuer: String,
        validFrom: Instant,
        validUntil: Instant,
        domains: List<String> = emptyList(),
        ipAddresses: List<String> = emptyList(),
        isCA: Boolean = false,
        issuerKeyPair: KeyPair? = null
    ): X509Certificate {
        val subjectDN = X500Name(subject)
        val issuerDN = X500Name(issuer)
        val serial = BigInteger.valueOf(System.currentTimeMillis())
        val signer = JcaContentSignerBuilder("SHA256withRSA").build(issuerKeyPair?.private ?: keyPair.private)

        val builder = JcaX509v3CertificateBuilder(
            issuerDN,
            serial,
            Date.from(validFrom),
            Date.from(validUntil),
            subjectDN,
            keyPair.public
        )

        // Add Basic Constraints extension
        builder.addExtension(
            Extension.basicConstraints,
            true,
            BasicConstraints(isCA)
        )

        // Add Subject Alternative Names
        if (domains.isNotEmpty() || ipAddresses.isNotEmpty()) {
            val sans = mutableListOf<GeneralName>()
            domains.forEach { sans.add(GeneralName(GeneralName.dNSName, it)) }
            ipAddresses.forEach { ip ->
                val ipBytes = InetAddress.getByName(ip).address
                println("Creating IP SAN for $ip with bytes: ${ipBytes.joinToString()}")
                val derOctetString = DEROctetString(ipBytes)
                println("DEROctetString: $derOctetString")
                val generalName = GeneralName(GeneralName.iPAddress, derOctetString)
                println("GeneralName: $generalName")
                sans.add(generalName)
            }
            builder.addExtension(
                Extension.subjectAlternativeName,
                false,
                GeneralNames(sans.toTypedArray())
            )
        }

        return JcaX509CertificateConverter().getCertificate(builder.build(signer))
    }

    private fun generateSelfSignedCertificate(
        keyPair: KeyPair,
        subject: String,
        validFrom: Instant,
        validUntil: Instant,
        domains: List<String> = emptyList(),
        ipAddresses: List<String> = emptyList(),
        isCA: Boolean = false
    ): X509Certificate {
        return generateCertificate(
            keyPair = keyPair,
            subject = subject,
            issuer = subject,
            validFrom = validFrom,
            validUntil = validUntil,
            domains = domains,
            ipAddresses = ipAddresses,
            isCA = isCA,
            issuerKeyPair = keyPair
        )
    }
} 